apps/webapp/app/v3/marqs.server.ts

```diff
@@ -0,0 +1,906 @@
+redis.call('SREM', concurrencyKey, messageId)
+
+-- Remove the message from the timeout queue
+redis.call('ZREM', visibilityQueue, messageId)
+
+-- Enqueue the message into the queue
+redis.call('ZADD', childQueueKey, messageScore, messageId)
+
+-- Rebalance the parent queue
+local earliestMessage = redis.call('ZRANGE', childQueueKey, 0, 0, 'WITHSCORES')
+if #earliestMessage == 0 then
+    redis.call('ZREM', parentQueueKey, childQueueName)
+else
+    redis.call('ZADD', parentQueueKey, earliestMessage[2], childQueueName)
+end
+`,
+    });
+
+    this.redis.defineCommand("heartbeatMessage", {
+      numberOfKeys: 1,
+      lua: `
+-- Keys: visibilityQueue
+local visibilityQueue = KEYS[1]
+
+-- Args: messageId, milliseconds, maxVisibilityTimeout
+local messageId = ARGV[1]
+local milliseconds = tonumber(ARGV[2])
+local maxVisibilityTimeout = tonumber(ARGV[3])
+
+-- Get the current visibility timeout
+local currentVisibilityTimeout = redis.call('ZSCORE', visibilityQueue, messageId)
+
+if currentVisibilityTimeout == nil then
+    return
+end
+
+-- Calculate the new visibility timeout
+local newVisibilityTimeout = math.min(currentVisibilityTimeout + milliseconds * 1000, maxVisibilityTimeout)
+
+-- Update the visibility timeout
+redis.call('ZADD', visibilityQueue, newVisibilityTimeout, messageId)
+      `,
+    });
+  }
+}
+
+declare module "ioredis" {
+  interface RedisCommander<Context> {
+    enqueueMessage(
+      queue: string,
+      parentQueue: string,
+      messageKey: string,
+      queueName: string,
+      messageId: string,
+      messageData: string,
+      messageScore: string,
+      callback?: Callback<void>
+    ): Result<void, Context>;
+
+    dequeueMessage(
+      childQueue: string,
+      parentQueue: string,
+      visibilityQueue: string,
+      concurrencyLimitKey: string,
+      currentConcurrencyKey: string,
+      childQueueName: string,
+      visibilityTimeout: string,
+      currentTime: string,
+      defaultConcurrencyLimit: string,
+      callback?: Callback<[string, string]>
+    ): Result<[string, string] | null, Context>;
+
+    acknowledgeMessage(
+      messageKey: string,
+      visibilityQueue: string,
+      concurrencyKey: string,
+      messageId: string,
+      callback?: Callback<void>
+    ): Result<void, Context>;
+
+    nackMessage(
+      messageKey: string,
+      childQueueKey: string,
+      parentQueueKey: string,
+      concurrencyKey: string,
+      visibilityQueue: string,
+      childQueueName: string,
+      messageId: string,
+      currentTime: string,
+      messageScore: string,
+      callback?: Callback<void>
+    ): Result<void, Context>;
+
+    heartbeatMessage(
+      visibilityQueue: string,
+      messageId: string,
+      milliseconds: string,
+      maxVisibilityTimeout: string,
+      callback?: Callback<void>
+    ): Result<void, Context>;
+  }
+}
+
+export const marqs = singleton("marqs", getMarQSClient);
+
+function getMarQSClient() {
+  if (env.REDIS_HOST && env.REDIS_PORT) {
```

Comment:
Should we log an error if these variables aren't set?

We could warn people that for v3 they will need to have Redis configured?